### ISSUE TITLE
FracdiffStat enormously slow

### ISSUE THREAD

**Issue by cTatu** [fracdiff/fracdiff#14]
Hi! I was trying to differentiate an Ethereum price dataset (attached below) but it is taking more than 1h and also uses more than 10 GB of memory. I don't mind the high memory usage, but it's taking too much time

```python
from fracdiff import FracdiffStat
import pandas as pd

df = pd.read_csv('ETHUSDT.csv.gz')
FracdiffStat().fit_transform(df.values)
```

[ETHUSDT.csv.gz](https://github.com/fracdiff/fracdiff/files/6406838/ETHUSDT.csv.gz)

**Comment by ghost** [comment]
Thanks for sharing that. Could you let me know the shape of the differentiated array?

**Comment by ghost** [comment]
Also, is `Fracdiff` slow too?

**Comment by cTatu** [comment]
`Fracdiff` is instantaneous! There are no issues.

The shape of the array is  `(229682, 5)`. Is historic stock candlestick data (OHLCV)

**Update**: It took like 9h to finish. Based on my observations, in order to find the minimum differentiation value to preserve the memory of the series `FracdiffStat` has to do a linear search. I saw that for every step it has to compute a linear regression and that's why it takes so long imo.

**Comment by ghost** [comment]
Thanks for letting me know!

I think the ADF test (`adfuller` in statsmodels) is the bottleneck.
As you point out, it performs linear regression which needs matrix multiplication and inversion.

Fractional differentiation is the linear order of n_samples (since it is essentially convolution) and order is binary-searched. These operations are optimized and unlikely to be the bottleneck.

**Comment by cTatu** [comment]
It would be super useful to have a stationary test that uses the GPU. Surprisingly I couldn't find any GPU implementation of ADF or KPSS. Is there something that I'm missing out, or I just didn't search thoroughly enough?

**Comment by ghost** [comment]
I think that would be awesome, but I've never heard of that kind of thing implemented, unfortunately. I guess https://github.com/rapidsai/cuml is somewhat promising.

**Comment by cTatu** [comment]
Thank you! There is in fact a gpu implementation of KPSS
https://github.com/rietmann-nv/cuml/pull/1/files

**Comment by cTatu** [comment]
Hi! I come to give a little update about alternative implementations of stationary tests.
I've tried cuml's implementation of KPSS, but it was horrendous instable. Luckily I found a Pytorch implementation of Dickey-Fuller Test that was much faster and used much less memory than the `statsmodels` one. I optimized further that code and I bring it to a x20 speedup

Here's the link if you're interested (I should make it a package):
https://github.com/cTatu/DFuller-Numpy
